chore: Autodetect cobc version and copybook dependencies

GitMensch opened the pull request:
* test: split compile from link + execute
* split download and unpack for server jar
* automatically set cobc-version

GitMensch commented:
I'm not sure yet where gitpod is going, but to test that: https://gitpod.io/?autostart=true#https://github.com/meyfa/CobolCraft/pull/311  (should start an environment, install necessary tools, compile cobolcraft (we may add init + version output #309 at its end) and runs its tests

meyfa commented:
I've never used Gitpod and am not really interested in adding it, however I do like working with Dev Containers (which is an open specification), and from what I can gather, Gitpod is fully compatible with those as well: https://www.gitpod.io/docs/flex/configuration/devcontainer

Would you be interested in changing this PR to add a devcontainer.json instead?

GitMensch commented:
I've never worked with devcontainers (and also haven't checked GitPod flex, which is necessary for that), so I'm likely not the right person to define those... If you want to I can drop the file from this PR (leaving only the Makefile) and then see if you find yourself some time to add a devcontainer definition someday.

meyfa commented on Makefile:
```diff
@@ -36,18 +37,20 @@ all: $(BIN) data
-$(SERVER_JAR_EXTRACTED):
+data: $(SERVER_JAR_EXTRACTED)
+
+data/server.jar:
```
This was intentionally NOT split, as it will ensure the .jar is re-downloaded whenever the version number changes.

I agree that splitting makes sense, but then we should add the version number into the downloaded filename (and extract it to a variable to avoid duplication).

meyfa commented on Makefile:
```diff
@@ -61,6 +64,9 @@ $(CPP_OBJECTS): $(CPP_SRC) $(CPP_HEADERS)
 
-test: $(CPY) $(TEST_SRC) $(OBJECTS) $(CPP_OBJECTS)
-	$(COBC) -x $(COBC_OPTS) -I $(CPY_DIR) -lstdc++ -lz -o $@ $(TEST_SRC) $(OBJECTS) $(CPP_OBJECTS)
+$(TEST_OBJECTS): tests/%.o: tests/%.cob $(CPY)
```
I would prefer if objects weren't placed next to sources. Besides looks, a practical reason is that `make clean` will only delete objects for which source files are still present, but if source files are renamed or removed, the objects will stay. With a separate directory, we can just delete the directory.

Perhaps place the test objects into a directory `test-out`? This would need to be added to `.dockerignore` and `.gitignore` as well, then.

GitMensch replied:
sounds good, I'll put it in out/test/ (the former directory is needed in any case)

meyfa reviewed (commented):
The automatic version detection is pretty nice! I've also left some comments about other aspects.

meyfa commented:
> If you want to I can drop the file from this PR (leaving only the Makefile) and then see if you find yourself some time to add a devcontainer definition someday.

Sure, that'd be fine :) Opened #312 to track this.

GitMensch commented:
This PR now includes "only" a heavy Makefile update; note that we can simplify it a lot if we remove the option to (re-)generate the dependency files on their own; if you want that I could drop it and do another push.

(and with GC 3.3+ the dependency generation gets better, but the 3.2 option should still be fine).

Side note: for my tests on GitPod I've opted to

```sh
wget https://cobolworx.com/pages/downloads/gnucobolcw_3.2-4_amd64.deb
sudo apt remove gnucobol
sudo apt autoremove
sudo apt install ./gnucobol*.deb
```

meyfa commented:
Looks interesting! I fear I won't get around to reviewing this today probably, as I'm currently working on the crafting game mechanic, and hope to get it somewhat finished today, but will review ASAP after that...

meyfa commented on Makefile:
```diff
@@ -8,59 +8,102 @@ MAIN_SRC = main.cob
-all: $(BIN) data
-
-clean:
+# GnuCOBOL >=3.2 added functions that drastically improve performance, query version (may one sub-digit) automatically
```
```suggestion
# GnuCOBOL >=3.2 added functions that drastically improve performance. Query the installed version (may have one sub-digit)
```

meyfa commented on Makefile:
```diff
@@ -8,59 +8,102 @@ MAIN_SRC = main.cob
-
-clean:
+# GnuCOBOL >=3.2 added functions that drastically improve performance, query version (may one sub-digit) automatically
+GCVERSION := $(shell $(COBC) --version | head -n1 | sed 's/.*\s\([0-9]\)\.\([0-9]\).*/\1\2/g' )
```
Override of `GCVERSION` should be removed from `.github/workflows/test.yml`, since it is redundant now

meyfa commented on Makefile:
```diff
@@ -8,59 +8,102 @@ MAIN_SRC = main.cob
 	rm -rf $(OBJECTS_DIR)
 	rm -f $(BIN)
-	rm -f $(TEST_BIN)
+	rm -f $(TEST_BIN) $(TEST_OBJECTS)
```
```suggestion
	rm -f $(TEST_BIN)
```

(test objects will already be removed due to `OBJECTS_DIR` being removed)

meyfa commented on Makefile:
```diff
@@ -8,59 +8,102 @@ MAIN_SRC = main.cob
+# Common compiler options, note: COB_FLAGS can be user-specified
+COBC_OPTS = --free -DGCVERSION=$(GCVERSION) -I $(CPY_DIR) $(COB_FLAGS)
+ifneq ($(DEPONLY),Y)
+COBC_OPTS += -g -O2 --debug -Wall -fnotrunc -fstatic-call
```
`-g` causes an enormous amount of files to be placed into the root directory (on GC 3.2), please remove it.

GitMensch replied:
I'll possibly check tomorrow on other workarounds (making the "out" directory the currdir, then execute cobc, comes to mind)

meyfa commented on Makefile:
```diff
@@ -8,59 +8,102 @@ MAIN_SRC = main.cob
+GLOBALDEPS =
+ifneq ($(DEPONLY),Y)
+# include dependency files for all existing object files
+-include $(OBJS:.o=.d)
```
```suggestion
-include $(OBJECTS:.o=.d)
```

meyfa reviewed (commented):
Very nice, seems to work great overall (with below suggestions applied). It's definitely great to not recompile everything when a copybook changes! Version detection also works great for 3.1.2, 3.2 and 3.3 and is definitely welcome since it takes the burden off of newcomers, in particular.

GitMensch commented:
I've dropped the "-g" as requested and adjusted to allow out-of tree-builds; this also allows for things like

```sh
mkdir build_debug
cd $_
make -f ../Makefile ROOT_DIR=.. COB_FLAGS=-g
```

resulting in binaries and full debug infos in a subfolder, which can then be used for GDB/vallgrind/perf analysis.

After we showcased that - what do you think of dropping the explicit dependency generation (separate commit), just to make the Makefile a little bit "cleaner"?
Anything else you'd like to see in this PR?

GitMensch commented:
Feel free to merge either both or only the first commit.

meyfa reviewed (approved):
LGTM, thank you!
